Source/Core/InputCommon/ControlReference/ControlReference.h

```diff
@@ -0,0 +1,79 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "InputCommon/ControlReference/ExpressionParser.h"
+#include "InputCommon/ControllerInterface/Device.h"
+
+// ControlReference
+//
+// These are what you create to actually use the inputs, InputReference or OutputReference.
+//
+// After being bound to devices and controls with UpdateReference,
+// each one can link to multiple devices and controls
+// when you change a ControlReference's expression,
+// you must use UpdateReference on it to rebind controls
+//
+
+class ControlReference
+{
+public:
+  static bool InputGateOn();
+
+  virtual ControlState State(const ControlState state = 0) = 0;
+  virtual ciface::Core::Device::Control* Detect(const unsigned int ms,
+                                                ciface::Core::Device* const device) = 0;
+
+  void UpdateReference(ciface::Core::DeviceContainer& devices,
+                       const ciface::Core::DeviceQualifier& default_device);
+
+  ControlState range;
+  std::string expression;
+  const bool is_input;
+  ciface::ExpressionParser::ExpressionParseStatus parse_error;
```

Comment:
Okay, I
- turned `is_input` into a virtual function; and
- renamed `parse_error` to `m_parse_status` and hid it behind a getter function.

`range` and `expression` are still public right now so they can be passed directly `IniFile::Section::Get`, which assigns to their pointers. I haven't prefixed them with `m_` for the same reason. Does that sound right?